### PR TITLE
feat(DTFS2-7466): add cancellation filter to portal deals

### PR DESCRIPTION
**Pull request by BethThomas141**
## Introduction :pencil2:
We want to be able to filter deals by those that are cancelled. 

## Resolution :heavy_check_mark:
- Add a 'cancelled' filter to the deals on portal
- Add E2E tests

## Miscellaneous :heavy_plus_sign:

image

### CODE REVIEW

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js [hunk 1]
```diff
@@ -24,183 +28,102 @@ context('Dashboard Deals filters - filter by status', () => {
...
+        checkboxSelector: statusCheckboxSelectors.cancelled,
+      },
+      {
+        status: ALL_STATUSES,
+        label: 'All-statuses',
+        checkboxSelector: statusCheckboxSelectors.all,
+      },
+    ];
+
+    statusesToTest.forEach(({ status, label, checkboxSelector }) => {
```
Comment (BethThomas141) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js:61]: I've refactored this test file by using a `forEach` to reduce repetition

File: e2e-tests/e2e-fixtures/constants.fixture.js [hunk 2]
```diff
@@ -33,6 +33,7 @@ export const DEAL_STATUS = {
   // statuses specific to the entire deal
   READY_FOR_APPROVAL: "Ready for Checker's approval",
   CHANGES_REQUIRED: "Further Maker's input required",
+  CANCELLED: 'Cancelled',
```
Comment (MarRobSoftwire) [inline e2e-tests/e2e-fixtures/constants.fixture.js:36]: Can this just be imported from libs/common?

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js [hunk 3]
```diff
@@ -24,183 +28,102 @@ context('Dashboard Deals filters - filter by status', () => {
...
+
+          const expectedText = `Remove this filter ${status}`;
+          cy.get(`[data-cy=main-container-selected-filter-${label}`).contains(expectedText);
+        });
+
+        if (status === ALL_STATUSES) {
+          it('renders all deals regardless of status', () => {
+            dashboardDeals.rows().should('have.length', ALL_DEALS.length);
+          });
+        } else {
```
Comment (MarRobSoftwire) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js:116]: I dont really like having an if statement here - could we use some functions to get around this whilst keeping the code DRY


```
describe.each(statusesExceptAllStatuses)('status', () => {
  withCommonTests();
  
  it(`renders only ${status} deals`, () => {
    // etc
  })
});

describe(ALL_STATUS, () => {
  withCommonTests();

  it('renders all deals regardless of status', () => {
    dashboardDeals.rows().should('have.length', ALL_DEALS.length);
  });
});

function withCommonTests(arguments) {
  it('does a thing', () => {
    // etc
  })
}
```

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filters/dashboard-deals-filters.spec.js [hunk 4]
```diff
@@ -152,6 +152,11 @@ context('Dashboard Deals filters', () => {
       dashboardDeals.filters.panel.form.status.rejectedByUKEF.checkbox().should('exist');
       dashboardDeals.filters.panel.form.status.rejectedByUKEF.checkbox().should('not.be.checked');
 
+      // cancelled
+      dashboardDeals.filters.panel.form.status.cancelled.label().contains(CONSTANTS.DEALS.DEAL_STATUS.CANCELLED);
+      dashboardDeals.filters.panel.form.status.cancelled.checkbox().should('exist');
+      dashboardDeals.filters.panel.form.status.cancelled.checkbox().should('not.be.checked');
```
Comment (MarRobSoftwire) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filters/dashboard-deals-filters.spec.js:158]: If this is behind a FF we might need to re-write it :(

File: portal/server/constants/status.js [hunk 5]
```diff
@@ -9,6 +9,7 @@ const DEAL = {
   UKEF_APPROVED_WITHOUT_CONDITIONS: 'Accepted by UKEF (without conditions)',
   UKEF_APPROVED_WITH_CONDITIONS: 'Accepted by UKEF (with conditions)',
   UKEF_REFUSED: 'Rejected by UKEF',
+  CANCELLED: 'Cancelled',
```
Comment (MarRobSoftwire) [inline portal/server/constants/status.js:12]: Can we use the commonised constants for this?

File: portal/server/controllers/dashboard/deals/template-filters.js [hunk 6]
```diff
@@ -80,6 +80,10 @@ const statusFilters = (submittedFilters) => {
       text: STATUS.DEAL.UKEF_REFUSED,
       value: STATUS.DEAL.UKEF_REFUSED,
     },
+    {
+      text: STATUS.DEAL.CANCELLED,
+      value: STATUS.DEAL.CANCELLED,
+    },
```
Comment (MarRobSoftwire) [inline portal/server/controllers/dashboard/deals/template-filters.js:86]: Should this be behind the FF?

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js (hunk 3 above)
Comment (ttbarnes) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js:116]: I'm finding both approaches hard to follow i.e:

```js
it('renders all deals regardless of status', () => {
it(`renders only ${status} deals`, () => {
```

I think we should be very specific, thinking something like this:

```js
  describe('when the `all statuses` checkbox/filter is applied', () => {
    it('should ....')
  });

  describe('when the `all statuses` checkbox/filter is NOT applied', () => {
    describe.each(`when the [ARRAY_OF_STATUSES] checkbox/filter is applied`, () => {
      it('should ....')
    });
  });

```

There are many combinations that we're probably not testing so would like to be as clear as possible even if it means more LOC

File: portal/server/constants/status.js (hunk 5 above)
Comment (BethThomas141) [reply portal/server/constants/status.js]: Good idea, now using👍

File: e2e-tests/e2e-fixtures/constants.fixture.js (hunk 2 above)
Comment (BethThomas141) [reply e2e-tests/e2e-fixtures/constants.fixture.js]: Good plan, now doing

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js (hunk 3 above)
Comment (BethThomas141) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js]: So given I've realised the cancellation filter needs to be behind a feature flag, I've actually just reverted this test file to be as it was, and made a new test file for the cancellation filter. 

So maybe out of scope of this ticket to make changes to this file anymore. 

Does that sound ok to you both?

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filters/dashboard-deals-filters.spec.js (hunk 4 above)
Comment (BethThomas141) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filters/dashboard-deals-filters.spec.js]: Great points, have redone the E2E tests

File: portal/server/controllers/dashboard/deals/template-filters.js (hunk 6 above)
Comment (BethThomas141) [reply portal/server/controllers/dashboard/deals/template-filters.js]: It should! Moved

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      dashboardDeals.filters.mainContainer.selectedFilters.statusCancelled().should('be.visible');
+
+      const expectedText = `Remove this filter ${TFM_DEAL_STAGE.CANCELLED}`;
+      dashboardDeals.filters.mainContainer.selectedFilters.statusCancelled().contains(expectedText);
+    });
+
+    it('renders only cancelled deals', () => {
+      dashboardDeals.rows().should('have.length', 1);
+
+      cy.assertText(dashboardDeals.row.status(BSS_DEAL_CANCELLED._id), TFM_DEAL_STAGE.CANCELLED);
```
Comment (MarRobSoftwire) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js:76]: Could you add a test that no others are rendered? I think there's a `row.length` property somewhere

Comment (MarRobSoftwire) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js:76]: On facilities you can do `dashboardFacilities.rows().should('have.length', EXPECTED_FACILITIES.length);`, hopefully theres equiv for deal

Comment (MarRobSoftwire) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js:76]: That or just add a `blah should not exist` test

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js (hunk 3 above)
Comment (MarRobSoftwire) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/filter-by/dashboard-deals-filter-by-status.spec.js]: That makes sense, leave as is for now :)

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js (hunk 7 above)
Comment (BethThomas141) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js:76]: There is already a `dashboardDeals.rows().should('have.length', 1);` which I think covers it!

Comment (BethThomas141) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js]: Have refactored anyway to make consistent with the other tests. Will mark as resolved!

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+const { dashboardDeals } = require('../../../../pages');
+const { dashboardFilters } = require('../../../../partials');
+const { BSS_DEAL_READY_FOR_CHECK, GEF_DEAL_DRAFT, BSS_DEAL_CANCELLED } = require('../../fixtures');
+
+const { BANK1_MAKER1, ADMIN } = MOCK_USERS;
+
+const filters = dashboardFilters;
+const statusCheckboxSelectors = dashboardDeals.filters.panel.form.status;
+
+context('Dashboard Deals Cancellation status filter - Deal cancellation enabled', () => {
```
Comment (ttbarnes) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filter-by-cancellation-status.spec.js:13]: Can we say "- X feature flag enabled" / "disabled" throughout please just so it's clear

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filters.spec.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  before(() => {
+    cy.login(BANK1_MAKER1);
+  });
+
+  describe('When deal cancellation is enabled on the dashboard deals page', () => {
+    it('displays the deal cancellation status filter', () => {
+      dashboardDeals.visit();
+      filters.showHideButton().click();
+
+      dashboardDeals.filters.panel.form.status.cancelled.label().contains(DEAL_STATUS.CANCELLED);
```
Comment (ttbarnes) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filters.spec.js:20]: Are we able to use `cy.assertText` here instead of `.contains`?

**Review by ttbarnes — COMMENTED** [review]
2x minor comments, otherwise LGTM, nice work

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filters.spec.js (hunk 9 above)
Comment (BethThomas141) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/deal-cancellation-enabled/dashboard-deals-filters.spec.js]: Good spot! have changed, thanks :)

**Review by ttbarnes — APPROVED** [review]
